### ISSUE TITLE
identify_package and locate_package introduced a data race regression in loading

### ISSUE THREAD

**Issue by vtjnash** [JuliaLang/julia#48511]
Regression introduced by #48247. Refs https://github.com/JuliaLang/julia/issues/48498

**Comment by KristofferC** [comment]
Could you be a bit clear what the race is? The caches for those are only used in `require` which is under the lock.

https://github.com/JuliaLang/julia/blob/c4fd8a48383235e818c44231d7507168c27f393b/base/loading.jl#L1568-L1570

**Comment by vtjnash** [comment]
I thought these are public API calls also. It is easiest to review the intent if function either asserts or (re)acquire the locks expected.

**Comment by KristofferC** [comment]
> I thought these are public API calls also.

Not really, they are not in the manual for example. People might use them for semi-interactive stuff but I don't think they would ever run concurrently with `require`.

**Comment by StefanKarpinski** [comment]
Can we just throw a lock around these?

**Comment by vtjnash** [comment]
Looking closer, I also found it strange that `parsed_toml` takes the lock as an optional argument, even though it is in fact mandatory for that argument to be `require_lock` @
https://github.com/JuliaLang/julia/blob/de7a7e768a068b2f3512c2b1ac71dfc74e500057/base/loading.jl#L273-L276

And later in the file, we seem to unconditionally erase the `LOADING_CACHE` content at every opportunity?
```
@lock require_lock LOADING_CACHE[] = LoadingCache()
....
@lock require_lock LOADING_CACHE[] = nothing
```

**Comment by KristofferC** [comment]
The `LOADING_CACHE` assumes many things do not change which can in fact change between `require` calls. It can be made more accurate so that it gets invalidated only when needed but this is what's there at the moment. I think the number of file accesses during a package loading is down 10x vs some old julia versions, partly due to this cache.

**Comment by omus** [comment]
@vtjnash I believe you fixed this in https://github.com/JuliaLang/julia/pull/59670?
